documentation/adr/0020-use-c-mex-api.md

```diff
@@ -0,0 +1,43 @@
+[<-previous](./0019-update-release-notes-every-pr.md) |
+next->
+
+# 20 - Use C Mex API (rather than C++)
+
+Date: 2021-Jan-20
+
+## Status
+
+Accepted
+
+## Context
+
+Matlab has two Mex APIs that can be used by C++ code:
+the [C API](https://www.mathworks.com/help/matlab/cc-mx-matrix-library.html)
+or the
+[C++ API](https://www.mathworks.com/help/matlab/cpp-mex-file-applications.html).
+The C++ API is relatively new and is therefore only compatible with Matlab releases
+[R2018a and later](https://www.mathworks.com/help/matlab/matlab_external/choosing-mex-applications.html#mw_d3e64706-faf9-486f-ab58-1860c63564d8).
+
+Throughout the Herbert and Horace codebases only the C API has been used.
+This is primarily due to the fact that Herbert and Horace pre-date the C++ API.
+The C++ API contains restrictions on directly accessing Matlab memory,
+which provided challenges when writing the Matlab object serializer in C++.
+
+TODO: do we have examples of where the C++ API did not work for our use-case?
+
+## Decision
+
+Herbert/Horace will continue to use the C API for new Mex functions.
+
+## Consequences
+
+- The C++ code we write will be compatible with all Matlab releases,
+  including those older than R2018a.
```

Comment:
Add additional distinction is between the "interleaved complex API" and the "separate complex API" - essentially whether complex matrices are stored in memory as separate real and imaginary matrices ("separate") or as a single matrix with elements being a complex data type ("interleave") - this was also introduced in R2018a. If we want to support Matlab versions before R2018a, we have to use the "separate complex API"... not sure if the Horace mex files uses complex numbers or not though, so this could be moot...

Reply:
Good point, I'll add that we'll have to continue using the separate complex API to maintain backwards compatibility where we use complex numbers. The only place I can see this being an issue is the C++ serializer (I think that's the only place we worry about complex numbers), would this be an issue @oerc0122?

Do you know if using the C++ Data API necessarily means using the newer "interleaved complex API"? I can't seem to find anything in the docs that says one way or the other. I had a quick check and I seem to be able to run a Mex file using the C++ Data API compiled with the `-R2017b` flag.

Reply:
There is absolutely no problem there and it's accounted for through the `MEX_SUPPORTS_INTERLEAVED_COMPLEX` or similar conditional compilation flag. I think the C++ API is always interleaved (due to native complex support in C++ not in C), but I may be wrong.

Reply:
Yes, I think the C++ Data API is always interleaved. For the separate complex API there is a C function (`mxGetPi` [compared with `mxGetPr` for the real part) to get a pointer to the imaginary matrix and the equivalent function in C++ does not exist.

I think the ADR looks fine now. 

(But, just a note for future... I think that internally within newer versions of Matlab complex numbers are stored in the interleaved format; so mex files which use the old API, whilst compatible with older versions of Matlab will suffer a performance penalty on newer versions of Matlab because Matlab will do an internal copy from the interleaved format to the separate format and vice versa on return from the mex file; This is one reason that for the update of SpinW we're targeting Matlab >= R2018a and using the C++ Data API.)